src/occupancy_constructor/scripts/occupancy_constructor.py

```diff
@@ -0,0 +1,147 @@
+        current_cone = current_cone.next
+
+        # Minimum has been met, allow loop closure
+        if len(loop) == min_loop_cones:
+            cones_remaining.append(first_cone)
+
+    cones_remaining.remove(first_cone)
+    return (loop, map_min, map_max, cones_remaining)
+
+
+def transform_cone_loop(loop, map_min, grid_resolution):
+    # Copy first point to the end, so it closes the loop
+    loop.append(Point(loop[0].x, loop[0].y))
+    loop_coords = []
+    for point in loop:
+        # Move all points to be relative to map_min
+        point.x -= map_min[0]
+        point.y -= map_min[1]
+        # Transfer scaled coords to a single list
+        loop_coords.append(point.x * grid_resolution)
+        loop_coords.append(point.y * grid_resolution)
+    return loop_coords
+
+
+def constructor_loop():
+    global cone_list, cone_list_dirty
+
+    # Get params for construction
+    min_loop_cones = rospy.get_param("occupancy_constructor/min_loop_cones")
+    grid_resolution = rospy.get_param("occupancy_constructor/grid_resolution")
+    wall_thickness = int(rospy.get_param("occupancy_constructor/wall_thickness") * grid_resolution)
+    grid_padding = rospy.get_param("occupancy_constructor/wall_thickness")
+
+    rate = rospy.Rate(rospy.get_param("occupancy_constructor/update_rate"))
+    pub = rospy.Publisher("/map", OccupancyGrid, queue_size=10)
+    while not rospy.is_shutdown():
+        if cone_list_dirty:
+            map_min = (cone_list[0].x, cone_list[0].y)
+            map_max = (cone_list[0].x, cone_list[0].y)
+            loop1, map_min, map_max, cones_remaining = traverse_cone_loop(
+                map_min, map_max, cone_list, min_loop_cones)
+            loop2, map_min, map_max, cones_remaining = traverse_cone_loop(
+                map_min, map_max, cones_remaining, min_loop_cones)
+
+            # Pad the map bounds by 1 meter
+            map_min = (map_min[0] - grid_padding, map_min[1] - grid_padding)
+            map_max = (map_max[0] + grid_padding, map_max[1] + grid_padding)
+            map_size = (map_max[0] - map_min[0], map_max[1] - map_min[1])
+
+            loop1_coords = transform_cone_loop(loop1, map_min, grid_resolution)
+            loop2_coords = transform_cone_loop(loop2, map_min, grid_resolution)
+
+            # Initialize OccupancyGrid
+            grid = OccupancyGrid()
+            grid.header.seq = 0
+            grid.header.frame_id = "map"
+            grid.info.resolution = 1 / grid_resolution
+            grid.info.width = math.ceil(map_size[0] * grid_resolution)
+            grid.info.height = math.ceil(map_size[1] * grid_resolution)
+            grid.info.origin.position.x = map_min[0]
+            grid.info.origin.position.y = map_min[1]
+            grid.info.origin.position.z = 0
+            grid.info.origin.orientation.x = 0
+            grid.info.origin.orientation.y = 0
+            grid.info.origin.orientation.z = 0
+            grid.info.origin.orientation.w = 1
```

Comment:
Are the default settings for `OccupancyGrid()` known? Because if so, we might be able to remove some unnecessary declarations.